### ISSUE TITLE
circular buffer overflow

### ISSUE THREAD

**Issue by timaschew** [grs/rhea#25]
Hi, I used your library in a hackathon for using Hono.
I'm using this code: https://github.com/bsinno/bcx2017/tree/master/Hono_Example_Clients/Hono-Node.js
I realized that I get sometimes these errors:

```
uncaught exception:  Error: circular buffer overflow: head=10 tail=10 size=2048 capacity=2048
    at CircularBuffer.push (/Users/timaschew/dev/iot-hackathon/hono-node-client/node_modules/rhea/lib/session.js:46:15)
    at Incoming.on_transfer (/Users/timaschew/dev/iot-hackathon/hono-node-client/node_modules/rhea/lib/session.js:303:29)
    at Session.on_transfer (/Users/timaschew/dev/iot-hackathon/hono-node-client/node_modules/rhea/lib/session.js:618:19)
    at Connection.(anonymous function) [as on_transfer] (/Users/timaschew/dev/iot-hackathon/hono-node-client/node_modules/rhea/lib/connection.js:642:30)
    at c.dispatch (/Users/timaschew/dev/iot-hackathon/hono-node-client/node_modules/rhea/lib/types.js:902:33)
    at Transport.read (/Users/timaschew/dev/iot-hackathon/hono-node-client/node_modules/rhea/lib/transport.js:95:36)
    at SaslClient.read (/Users/timaschew/dev/iot-hackathon/hono-node-client/node_modules/rhea/lib/sasl.js:243:26)
    at Connection.input (/Users/timaschew/dev/iot-hackathon/hono-node-client/node_modules/rhea/lib/connection.js:420:35)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:191:7)
```

**Comment by grs** [comment]
Thanks for raising an issue! I think I see the issue and will attempt to reproduce and fix.

In the meantime, as a workaround, I would suggest modifying the eventReceiverOptions to have rcv_settle_mode set to 0 (i.e. first, which is the default). That gives at-least-once delivery. The only value in setting it to second would be if you were going to resume the link on disconnect and resolve unsettled state. As the example doesn't do that (and to tell the truth, I have never tested that yet with rhea as most brokers I have tested against don't yet support it either), there is not much gained by expecting the sender to settle.

**Comment by ps2goat** [comment]
@grs - any update on this? I just recently got it, also, during my throughput testing.  It looks like if you try to send a lot of messages at once for a single `Sender`, faster than the broker can acknowledge them, then the buffer fills up and throws this error.

**Comment by grs** [comment]
Are you checking Sender.sendable() when sending?

**Comment by amarzavery** [comment]
That is true. You need to check for sender.sendable() while sending messages. You can take a look at what i am doing for Azure event hubs over [here](https://github.com/Azure/azure-event-hubs-node/blob/master/client/lib/eventHubSender.ts#L180)

**Comment by ps2goat** [comment]
Thanks @grs and @amarzavery -- it is working now.  But was the original issue fixed? I'm curious, as it's been over a year.

**Comment by amarzavery** [comment]
Depends on how you look at it. I don't think it is an issue. Rhea tells.you that you are doing something wrong. Please settle some of your messages if you are a receiver or make sure that you send messages at a plausible rate if you are a sender. There is no point keeping so many messages in the circular buffer/ memory.

**Comment by ps2goat** [comment]
@amarzavery - I didn't mean my issue.  I meant the first post that @grs responded to last year.  I want to know if it is an actual issue that needs to be fixed, if it's been fixed already, etc.

Thanks!

**Comment by grs** [comment]
@ps2goat The original issue affects receivers, not senders. It only affects the 'receiver settles last' mode (which is off by default) used in conjunction with an automatically moved credit window (which is on by default). It still needs fixed.

**Comment by Abhishek150995** [comment]
Hi
**context.session.outgoing.deliveries.capacity** has the value of max size of Circular Buffer
Is there a way to know if the max size if about to reach so that I can Stop publishing?

What does **context.session.outgoing.deliveries.size** , **.tail, .head** signify?

**Comment by ps2goat** [comment]
@Abhishek150995 - I haven't looked at this in a while, the maintainers can correct me if I'm wrong.

The circular buffer is like a queue of messages to send. `context.session.outgoing.deliveries.capacity` is how many messages can be stored at once (how big the buffer can grow to), and `context.session.outgoing.deliveries.size` is how big the buffer has grown so far. The buffer is lazily created, so we have an empty array to start with (size = 0), but it will grow to the max capacity over time. `size` tracks the number of messages pending send.

`.tail` is the end of the current queue (messages to be sent), and `.head` is the start of the current queue. Let's say I have a buffer with a capacity of 4. (It looks like this is currently set to 2048, and not really exposed for easy configuration). Note that in my examples, `-` represents a null item and a number represents a message placed on the buffer/queue.

The buffer starts as an empty array.
> `[ ]`

Now I fill this with 2 items. The array grows, but is not yet at its full capacity of 4.
> `[ 1 2 ]`
`.head` is 1, `.tail` is 2. `.size` is 2. As messages are sent, the size is reduced.

When those are processed, their spot is opened up in the buffer (`-` in my illustration). I believe the process tracks the next index to use based on the last inserted value's location, so it would then look like this:
> `[ - - 3 ]`
`.head` and `.tail` would both point at 2 (0-based index). Since the capacity is 4 in my example, the third index was added to the buffer. `size` is one here, as there is one message to be sent.

Insert 3 more (4, 5, 6), and it would start to wrap around like this:
> `[ 5 6 - 4 ]`
The fourth buffer spot was added, but then the capacity was met, and the next messages start inserting at the front of the buffer again. `size` here is 3, as there are 3 messages to send.

Once you've sent enough messages to have created a full buffer, and all messages were sent, you'd have an array of 4 empty spots (capacity), a `size` of zero, and `head` and `tail` would point to the same index, waiting for a new message to be added to the buffer.
> `[ - - - - ]`

**Comment by grs** [comment]
> Is there a way to know if the max size if about to reach so that I can Stop publishing?

`context.session.outgoing.available()` will give the number of messages that can still be sent
